porcupine/plugins/run/no_terminal.py

```diff
@@ -122,6 +123,7 @@ def _thread_target(self, command: str, env: dict[str, str]) -> None:
         else:
             self._queue.put(("error", f"The process failed with status {status}."))
         self._queue.put(("end", ""))
+        self.running = False
```

Comment:
I don't like this attribute: if you forget to set it in one of the many places, we suddenly have a bug that happens rarely and is hard to notice when it does happen. Maybe instead make a method (or property) that checks whether `self._shell_process` is non-None and alive? This way we don't store the same information in two places, in the process object and in a separate boolean.